enhancements/20241211-upgrade-enhancement.md

```diff
@@ -0,0 +1,997 @@
+
+The promotion of Upgrade Manager is to eliminate the two-version fix cycle paradox. Even further, its duty is deliberately stretched to cover most of the Harvester Upgrade mechanics to serve the same purpose. Besides, upgrade-related logic is only needed during upgrades and is rarely used at other times, so it does not need to exist all the time. That's the decisive reason not to deploy Upgrade Manager as a permanent deployment.
+
+The introduction of Upgrade Manager is the most crucial part of the Harvester Upgrade V2 enhancement. It not only separates the upgrade reconcile loops from the bulky Harvester controller manager, but also changes how node-upgrade used to work in Upgrade V1. It encompasses two stages to give us enough time to incorporate all the changes and make them take effect.
+
+#### Stage 1 - Transition Period (version N-1 to N upgrades, where N is the debut minor version of Harvester Upgrade V2)
+
+The inner workings of Upgrade Manager in this stage will be essentially the same as before; the significant difference is that it is built and packaged separately from the primary Harvester artifact. It will also have its own Deployment in contrast to the main Harvester controller manager Deployment. The main concerns will be:
+
+- How Upgrade Manager is bootstrapped
+- How Upgrade Manager is torn down
+
+The decision is straightforward: since Upgrade Shim initiates Upgrade Manager, to ensure workflow symmetry, it should also be performed by Upgrade Shim for teardown.
+
+> [!NOTE]
+> The N-to-N upgrades, including patch version upgrades, are deemed stage 1. For instance, the following upgrade paths are the case:
+>
+> - v1.7.0 -> v1.7.0
+> - v1.7.0 -> v1.7.1
+> - v1.7.1 -> v1.7.1
+
+#### Stage 2 - Complete Upgrade V2 (version N to N+1 upgrades, where N is the debut minor version of Harvester Upgrade V2)
+
+Harvester Upgrade V2 is fully unlocked in this stage. The cluster-upgrade phase largely remains the same. In the node-upgrade phase, Rancher v2prov will be dropped, and Harvester Upgrade will no longer rely on the embedded Rancher and Rancher System Agent on each node. Instead, System Upgrade Controller becomes a crucial companion.
+
+The highlights in the original node-upgrade phase are:
+
+1. User workloads, i.e., VMs, evacuation, corresponds to the **pre-drain** stage
+1. Node drain and RKE2 upgrade, corresponds to the **drain** stage
+1. OS upgrade, corresponds to the **post-drain** stage
+
+Since the hook mechanism in Rancher v2prov is no longer usable, Upgrade Manager relies on Plan CRs backed by System Upgrade Controller to execute node-upgrade tasks. One of the hidden advantages of this change is that it allows the operating system upgrade to be separated from the overall upgrade, thus enabling true zero-downtime (for nodes) upgrades. Another one is that it unblocks the way for users to have granular control over the node-upgrade order through node selectors. The new node-upgrade phase looks like the following:
```

Comment:
I am not really sure about that. If we cannot move to SUC plan, we can only use hook mechanism like current upgrade flow, right?

Reply:
Yes, that's the backup plan if the SUC path is not feasible.